### ISSUE TITLE
False negative for `unused_mut` when writing to field through reference

### ISSUE THREAD

**Issue by lukas-code** [rust-lang/rust#110849]
### Code

I tried [this code](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=b00a8e7868624d17841d8c23c3137217):

```rust
pub struct Foo(i32);

pub fn foo(mut x: &mut Foo) {
    x.0 = 1;
}
```

I expected to see this: The compiler should tell me to remove the `mut`, since it isn't necessary:
```text
warning: variable does not need to be mutable
 --> src/lib.rs:3:12
  |
3 | pub fn foo(mut x: &mut Foo) {
  |            ----^
  |            |
  |            help: remove this `mut`
  |
  = note: `#[warn(unused_mut)]` on by default
```

Instead this happened: The code compiles without a lint.

The lint triggers correctly with an explicit reborrow: `(&mut *x).0 = 1;`

### Meta

playground 1.71.0-nightly (2023-04-25 458d4dae845ec155b285)

@rustbot label A-diagnostics A-lint
